specs/capella/light-client/fork.md

```diff
@@ -0,0 +1,92 @@
+# Capella Light Client -- Fork Logic
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+  - [Upgrading light client data](#upgrading-light-client-data)
+  - [Upgrading the store](#upgrading-the-store)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This document describes how to upgrade existing light client objects based on the [Altair specification](../../altair/light-client/sync-protocol.md) to Capella. This is necessary when processing pre-Capella data with a post-Capella `LightClientStore`. Note that the data being exchanged over the network protocols uses the original format.
+
+### Upgrading light client data
+
+A Capella `LightClientStore` can still process earlier light client data. In order to do so, that pre-Capella data needs to be locally upgraded to Capella before processing.
+
+```python
+def upgrade_lc_header_to_capella(pre: BeaconBlockHeader) -> LightClientHeader:
+    return LightClientHeader(
+        beacon=pre,
+    )
+```
+
+```python
+def upgrade_lc_bootstrap_to_capella(pre: altair.LightClientBootstrap) -> LightClientBootstrap:
```

Comment:
Although `altair.LightClientBootstrap` SSZ fields are identical `bellatrix.LightClientBootstrap`, I still think it should be `bellatrix.LightClientBootstrap` here for the fork transition. That is, we need `bellatrix/light-client/fork.md` and `upgrade_*_to_bellatrix` helpers.

Moreover, should we make `bellatrix.LightClientHeader` include `ExecutionPayloadHeader` and the Merkle proof? It seems more natural since the `ExecutionPayloadHeader` field was first introduced in Bellatrix. However, (1) it's less useful since Capella is coming soon, and (2) we are still discussing #3078. So I'd say it's fine to keep  `bellatrix.LightClientHeader` as simple as `altair.LightClientHeader`.

Reply:
Hmm, not sure about the nop `upgrade_*_to_bellatrix` — this would then be required for any future fork, regardless of whether there are any real updates.

Kinda like it with `altair` for readability (in practice, implementations will reuse the Altair definition for Bellatrix and not duplicate all the code with every fork). Agree though, that it is a bit hacky as proposed.

Anyhow, if you insist, can update it (it works well as-is in pytests though).

About changing Bellatrix data: that one is already being cached in DBs and gossiped without the `ExecutionPayloadHeader`. Changing that would be quite tricky (and in case of gossip may even need a version bump on those topics to avoid disconnecting peers on the previous version). Would therefore leave Bellatrix as is, also considering Capella is in the near future.

Reply:
About `upgrade_*_to_bellatrix`, those are actually not needed. These `upgrade` helpers are only used when obtaining Altair formatted data and trying to feed it to an already-upgraded `LightClientStore`.

In the `p2p-interface.md` files you can see that it is not possible to obtain a `bellatrix.LightClient*` object.

| `fork_version`                                         | Response chunk SSZ type               |
| ------------------------------------------------------ | ------------------------------------- |
| `GENESIS_FORK_VERSION`                                 | n/a                                   |
| `ALTAIR_FORK_VERSION` through `BELLATRIX_FORK_VERSION` | `altair.LightClientUpdate`            |
| `CAPELLA_FORK_VERSION`                                 | `capella.LightClientUpdate`           |
| `EIP4844_FORK_VERSION` and later                       | `eip4844.LightClientUpdate`           |

Reply:
And, with this, it is actually correct to have the upgrade convert directly from `altair.LightClient*` to `capella.LightClient*`.

Reply:
(as in, there is no way to actually obtain a `bellatrix.LightClientUpdate` in practice, so no upgrade facility is needed).

Reply:
Now I agree on **no** need to implement `upgrade_*_to_bellatrix` and the whole `bellatrix/light-client/fork.md` file. 👍

But I still think it's good to use the "previous fork" (bellatrix) in `upgrade_*_to_capella` in specs.
- In markdown specs, `bellatrix.X` means the class/method version we have in Bellatrix.
- In pyspec, we automatically "copy" the code from the previous definition. It's like an inheritance. So `bellatrix.LightClientUpdate` exists in pyspec usage, although it is just as same as `altair.LightClientUpdate` definition.
- If we can *always* use the "previous fork" in the fork transition specs, I think it would be fewer potential bugs. i.e., we don't have to worry about which fork to use if we have this pattern.

If client implementations actually skip the "unchanged-phase" objects, it's also acceptable since the results are the same.

Reply:
Sounds good, will change the function signature to `bellatrix.X` but omit the `upgrade_*_to_bellatrix` stub.